File: python-backend-backup/drive_service.py
"""
drive_service.py — Fungsi-fungsi untuk berinteraksi dengan Google Drive API.
Termasuk: buat service, cek kuota, list file, upload, download, delete, rename.
"""

import io
import tempfile
from google.oauth2.credentials import Credentials
from googleapiclient.discovery import build
from googleapiclient.http import MediaIoBaseUpload, MediaIoBaseDownload


def get_drive_service(account):
    """Buat Google Drive API service dari kredensial akun yang tersimpan."""
    creds = Credentials(
        token=account.access_token,
        refresh_token=account.refresh_token,
        token_uri=account.token_uri,
        client_id=account.client_id,
        client_secret=account.client_secret,
        scopes=['https://www.googleapis.com/auth/drive']
    )
    service = build('drive', 'v3', credentials=creds, cache_discovery=False)
    return service, creds


def get_storage_quota(account):
    """Ambil informasi kuota penyimpanan untuk satu akun."""
    service, creds = get_drive_service(account)
    about = service.about().get(fields='storageQuota,user').execute()
    quota = about.get('storageQuota', {})
    user = about.get('user', {})
    return {
        'total': int(quota.get('limit', 0)),
        'used': int(quota.get('usage', 0)),
        'email': user.get('emailAddress', ''),
        'display_name': user.get('displayName', '')
    }


def list_files(account, folder_id='root', page_size=100):
    """Ambil daftar file dan folder dari folder tertentu di Drive."""
    service, creds = get_drive_service(account)
    query = f"'{folder_id}' in parents and trashed = false"
    results = service.files().list(
        q=query,
        pageSize=page_size,
        fields="nextPageToken, files(id, name, mimeType, size, modifiedTime, parents, iconLink, thumbnailLink, webViewLink)",
        orderBy="folder,name"
    ).execute()
    files = results.get('files', [])
    # Tambahkan info akun ke setiap file
    for f in files:
        f['account_id'] = account.id
        f['account_email'] = account.email
    return files


def upload_file(account, file_stream, filename, mime_type, folder_id='root', progress_callback=None):
    """Upload file ke Google Drive."""
    service, creds = get_drive_service(account)
    file_metadata = {
        'name': filename,
        'parents': [folder_id]
    }
    # Gunakan chunksize 10MB (harus kelipatan 256KB) untuk efisiensi memori
    media = MediaIoBaseUpload(file_stream, mimetype=mime_type, chunksize=10*1024*1024, resumable=True)
    request = service.files().create(
        body=file_metadata,
        media_body=media,
        fields='id, name, size'
    )
    
    response = None
    while response is None:
        status, response = request.next_chunk()
        if status and progress_callback:
            try:
                progress_callback(status.resumable_progress, status.total_size)
            except Exception as e:
                print(f"Error in progress callback: {e}")
                
    return response


def delete_file(account, file_id):
    """Hapus file dari Google Drive."""
    service, creds = get_drive_service(account)
    service.files().delete(fileId=file_id).execute()


def rename_file(account, file_id, new_name):
    """Rename file di Google Drive."""
    service, creds = get_drive_service(account)
    file = service.files().update(
        fileId=file_id,
        body={'name': new_name},
        fields='id, name'
    ).execute()
    return file


def download_file(account, file_id):
    """Download file dari Google Drive. Mendukung export Google Docs."""
    service, creds = get_drive_service(account)
    # Ambil metadata file dulu
    file_meta = service.files().get(fileId=file_id, fields='name, mimeType').execute()

    # Google Docs types perlu di-export ke format standar
    google_export_map = {
        'application/vnd.google-apps.document': ('application/pdf', '.pdf'),
        'application/vnd.google-apps.spreadsheet': (
            'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet', '.xlsx'
        ),
        'application/vnd.google-apps.presentation': (
            'application/vnd.openxmlformats-officedocument.presentationml.presentation', '.pptx'
        ),
    }

    mime_type = file_meta.get('mimeType', '')
    filename = file_meta.get('name', 'download')

    if mime_type in google_export_map:
        export_mime, ext = google_export_map[mime_type]
        request = service.files().export_media(fileId=file_id, mimeType=export_mime)
        if not filename.endswith(ext):
            filename += ext
    else:
        request = service.files().get_media(fileId=file_id)

    # Download ke temporary file di disk (bukan memori) agar hemat RAM
    fh = tempfile.TemporaryFile()
    downloader = MediaIoBaseDownload(fh, request)
    done = False
    while not done:
        status, done = downloader.next_chunk()
    fh.seek(0)

    return fh, filename, mime_type


def fetch_all_files(account):
    """Ambil seluruh file dan folder dari satu akun Google Drive secara massal."""
    service, creds = get_drive_service(account)
    query = "trashed = false"
    
    files = []
    page_token = None
    while True:
        results = service.files().list(
            q=query,
            pageSize=1000,
            fields="nextPageToken, files(id, name, mimeType, size, modifiedTime, parents, starred, shared)",
            pageToken=page_token
        ).execute()
        files.extend(results.get('files', []))
        page_token = results.get('nextPageToken')
        if not page_token:
            break
            
    return files


def create_folder(account, folder_name, parent_id='root'):
    """Buat folder baru di Google Drive."""
    service, creds = get_drive_service(account)
    file_metadata = {
        'name': folder_name,
        'mimeType': 'application/vnd.google-apps.folder',
        'parents': [parent_id]
    }
    folder = service.files().create(
        body=file_metadata,
        fields='id, name'
    ).execute()
    return folder


def toggle_starred(account, file_id, starred):
    """Toggle status bintang file di Google Drive."""
    service, creds = get_drive_service(account)
    body = {'starred': starred}
    service.files().update(fileId=file_id, body=body, fields='id, starred').execute()


def toggle_shared(account, file_id, shared):
    """Toggle status berbagi file di Google Drive (anyone reader permission)."""
    service, creds = get_drive_service(account)
    if shared:
        permission = {
            'type': 'anyone',
            'role': 'reader'
        }
        service.permissions().create(fileId=file_id, body=permission, fields='id').execute()
    else:
        # Cari izin 'anyone' dan hapus
        permissions = service.permissions().list(fileId=file_id, fields='permissions(id, type)').execute().get('permissions', [])
        for perm in permissions:
            if perm.get('type') == 'anyone':
                service.permissions().delete(fileId=file_id, permissionId=perm.get('id')).execute()

File: driveService.js
import { google } from 'googleapis';
import fs from 'fs';
import path from 'path';
import os from 'os';

/**
 * Get Google Drive API service client for an account.
 * Automatically refreshes access token using refresh token and saves it.
 */
export function getDriveService(account) {
  const oauth2Client = new google.auth.OAuth2(
    account.client_id,
    account.client_secret,
    account.token_uri
  );
  
  oauth2Client.setCredentials({
    access_token: account.access_token,
    refresh_token: account.refresh_token
  });

  // Automatically save new access tokens when they are refreshed
  oauth2Client.on('tokens', (tokens) => {
    if (tokens.access_token) {
      account.access_token = tokens.access_token;
      account.save().catch(err => {
        console.error(`Failed to auto-update access token for ${account.email}:`, err);
      });
    }
  });

  const drive = google.drive({ version: 'v3', auth: oauth2Client });
  return { drive, oauth2Client };
}

/**
 * Get storage quota info for a single account.
 */
export async function getStorageQuota(account) {
  const { drive } = getDriveService(account);
  const res = await drive.about.get({ fields: 'storageQuota,user' });
  const quota = res.data.storageQuota || {};
  const user = res.data.user || {};
  return {
    total: parseInt(quota.limit || '0', 10),
    used: parseInt(quota.usage || '0', 10),
    email: user.emailAddress || '',
    display_name: user.displayName || ''
  };
}

/**
 * Get list of files/folders in a folder.
 */
export async function listFiles(account, folderId = 'root', pageSize = 100) {
  const { drive } = getDriveService(account);
  const query = `'${folderId}' in parents and trashed = false`;
  const res = await drive.files.list({
    q: query,
    pageSize: pageSize,
    fields: 'nextPageToken, files(id, name, mimeType, size, modifiedTime, parents, iconLink, thumbnailLink, webViewLink)',
    orderBy: 'folder,name'
  });
  
  const files = res.data.files || [];
  return files.map(f => ({
    ...f,
    account_id: account.id,
    account_email: account.email
  }));
}

/**
 * Upload a file stream to Google Drive.
 */
export async function uploadFile(account, fileStream, filename, mimeType, folderId = 'root', fileSize = null, progressCallback = null) {
  const { drive } = getDriveService(account);
  
  const requestBody = {
    name: filename,
    parents: [folderId]
  };

  const media = {
    mimeType: mimeType,
    body: fileStream
  };

  const uploadOptions = {};
  if (fileSize && progressCallback) {
    uploadOptions.onUploadProgress = (evt) => {
      progressCallback(evt.bytesRead, fileSize);
    };
  }

  const res = await drive.files.create({
    requestBody,
    media,
    fields: 'id, name, size'
  }, uploadOptions);

  return res.data;
}

/**
 * Delete a file/folder from Google Drive.
 */
export async function deleteFile(account, fileId) {
  const { drive } = getDriveService(account);
  await drive.files.delete({ fileId });
}

/**
 * Rename a file/folder in Google Drive.
 */
export async function renameFile(account, fileId, newName) {
  const { drive } = getDriveService(account);
  const res = await drive.files.update({
    fileId,
    requestBody: { name: newName },
    fields: 'id, name'
  });
  return res.data;
}

/**
 * Download a file from Google Drive (handles Google Docs exports and streaming).
 */
export async function downloadFile(account, fileId) {
  const { drive } = getDriveService(account);
  
  // Get metadata first
  const metaRes = await drive.files.get({ fileId, fields: 'name, mimeType' });
  const mimeType = metaRes.data.mimeType || '';
  let filename = metaRes.data.name || 'download';
  
  const googleExportMap = {
    'application/vnd.google-apps.document': { mime: 'application/pdf', ext: '.pdf' },
    'application/vnd.google-apps.spreadsheet': { mime: 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet', ext: '.xlsx' },
    'application/vnd.google-apps.presentation': { mime: 'application/vnd.openxmlformats-officedocument.presentationml.presentation', ext: '.pptx' }
  };

  let downloadRes;
  if (googleExportMap[mimeType]) {
    const exportMeta = googleExportMap[mimeType];
    downloadRes = await drive.files.export({
      fileId,
      mimeType: exportMeta.mime
    }, { responseType: 'stream' });
    
    if (!filename.endsWith(exportMeta.ext)) {
      filename += exportMeta.ext;
    }
  } else {
    downloadRes = await drive.files.get({
      fileId,
      alt: 'media'
    }, { responseType: 'stream' });
  }

  // Create temporary file path
  const tempPath = path.join(os.tmpdir(), `gabungin_${Date.now()}_${Math.random().toString(36).substring(2)}`);
  const writeStream = fs.createWriteStream(tempPath);
  
  await new Promise((resolve, reject) => {
    downloadRes.data.pipe(writeStream);
    downloadRes.data.on('error', reject);
    writeStream.on('finish', resolve);
    writeStream.on('error', reject);
  });

  const readStream = fs.createReadStream(tempPath);
  const cleanUp = () => {
    fs.unlink(tempPath, (err) => {
      if (err) console.error(`Failed to delete temporary file: ${tempPath}`, err);
    });
  };

  return { readStream, filename, mimeType, cleanUp };
}

/**
 * Fetch all files and folders from Google Drive recursively (bulk request).
 */
export async function fetchAllFiles(account) {
  const { drive } = getDriveService(account);
  const query = 'trashed = false';
  const files = [];
  let pageToken = null;
  
  do {
    const res = await drive.files.list({
      q: query,
      pageSize: 1000,
      fields: 'nextPageToken, files(id, name, mimeType, size, modifiedTime, parents, starred, shared)',
      pageToken: pageToken
    });
    
    files.push(...(res.data.files || []));
    pageToken = res.data.nextPageToken;
  } while (pageToken);
  
  return files;
}

/**
 * Create a new folder.
 */
export async function createFolder(account, folderName, parentId = 'root') {
  const { drive } = getDriveService(account);
  const res = await drive.files.create({
    requestBody: {
      name: folderName,
      mimeType: 'application/vnd.google-apps.folder',
      parents: [parentId]
    },
    fields: 'id, name'
  });
  return res.data;
}

/**
 * Star / unstar a file.
 */
export async function toggleStarred(account, fileId, starred) {
  const { drive } = getDriveService(account);
  await drive.files.update({
    fileId,
    requestBody: { starred: starred },
    fields: 'id, starred'
  });
}

/**
 * Toggle file sharing (public reader access).
 */
export async function toggleShared(account, fileId, shared) {
  const { drive } = getDriveService(account);
  if (shared) {
    await drive.permissions.create({
      fileId,
      requestBody: {
        type: 'anyone',
        role: 'reader'
      },
      fields: 'id'
    });
  } else {
    const res = await drive.permissions.list({ fileId, fields: 'permissions(id, type)' });
    const permissions = res.data.permissions || [];
    for (const perm of permissions) {
      if (perm.type === 'anyone') {
        await drive.permissions.delete({ fileId, permissionId: perm.id });
      }
    }
  }
}
